ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,273 @@
+example:
+
+```ruby
+class SumActivity < Temporal::Activity
+  input Integer, Integer
+  output Integer
+
+  def execute(a, b)
+    a + b
+  end
+end
+```
+
+We will support most basic Ruby types (e.g. `Integer`, `Float`, `String`, `Symbol`, `Boolean`,
+`Array`, `Hash`, etc). This will allow the SDK to perform a runtime check of the arguments and
+return values to ensure they match the specified types.
+
+We will also leverage this system to allow SDK users to specify their own types:
+
+```ruby
+class MyInput < Struct.new(:a, :b)
+  def self.load(hash)
+    new(hash['a'], hash['b'])
+  end
+
+  def dump
+    { 'a' => a, 'b' => b } # explicit for illustration purposes, can use #to_h instead
+  end
+end
+
+class MyOutput < Struct.new(:sum)
+  def self.load(sum)
+    new(sum.to_i)
+  end
+
+  def dump
+    sum.to_s # string used for illustration purposes, can use Integer directly instead
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  input MyInput
+  output MyOutput
+
+  def execute(input)
+    MyOutput.new(input.a + input.b)
+  end
+end
+```
+
+Custom types are expected to implement the following interface:
+
+```ruby
+interface _CustomType[U]
+  # Take a primitive type and return an instance of the implementing class
+  def self.load: (U) -> instance
+
+  # Returns a trivially serializable type representing the custom object
+  def dump: -> U
+end
+```
+
+The same exact concept applies to signals and queries:
+
+```ruby
+class MyWorkflow < Temporal::Workflow
+  signal :add, input: Integer
+  query :get_sum, output: MyOutput
+
+  def execute
+    ...
+  end
+end
+```
+
+Notes:
+
+- This feature is 100% optional. Without these definitions the output from converters will be used
+  directly as arguments and return values
+- The full potential of this solution is obviously limited to a single language SDKs
+- When sending a custom input type to another SDK, workflow/activity definitions (without the
+  implementation) can still be leveraged
+- A custom output type sent to another SDK can still be `dump`ed into a matching JSON or other
+  serializable structure
+- The same approaches apply to inputs and outputs received from other SDKs
+
+
+## Payload handling
+
+Ruby SDK will use the model proposed [here](https://docs.temporal.io/docs/concepts/what-is-a-data-converter/)
+to distinguish between payload converters and codecs.
+
+The main difference is that a Converter translates between any Ruby object and a Payload, while
+a Codec is a lower level construct that encodes/decodes from a Payloads protobuf to a Payloads
+protobuf. The former is intended to support specific serialisation formats, while the latter is for
+modifying a binary that is transmitted over the wire.
+
+
+### Converters
+
+A payload converter is expected to implement the following interface:
+
+```ruby
+interface _PayloadConverter[U]
+  # Returns whether or not it can decode a given Payload
+  def can_decode?: (Temporal::Payload) -> Bool
```

Comment:
I'm fine with this approach replacing the "EncodingDataConveter" concept in other SDKs.

Reply:
I don't understand this. Where do I set the encoding string to be used in a dictionary? We shouldn't call `can_decode` in order for trying to decode when we should be using the encoding to determine that instead, right

Also, why does a high-level payload converter need to know whether it can decode. IMO a data converter has a single payload converter and that's it. If the payload converter wants to delegate, it can, but the high-level interfaces is just a simple to/from.

Arguably this would be called `can_convert` or similar since decoding is not the common word we use on converters.

Reply:
Updated